### DISCUSSION TITLE
RelationManager with Pivot and Select Multi

### DISCUSSION THREAD

**Discussion by warsemann** [filamentphp/filament#5834] [Help]
Hi, it s possible to add multi select for pivot column in relation manager ?
Example :
![image](https://user-images.githubusercontent.com/77194719/222524553-9e7dd46a-2127-4383-9dcb-bc98edf428b0.png)

But when i attach i have an error : "Array to string conversion"

Thanks !

**Comment by danharrin** [comment]
You need a pivot model to cast the attribute to an array

**Reply by warsemann** [reply]
It's OK thanks, it s possible to attach multiple in AttachAction ?

**Reply by danharrin** [reply]
You can probably chain ->multiple() onto the getRecordSelect() ?

**Reply by warsemann** [reply]
Yes your the best ;)

I think it s not possible to do what i want, i have a table with three relation :
![image](https://user-images.githubusercontent.com/77194719/222702653-736231f7-4ae6-4bb6-b783-e5361cc661c5.png)

I want to attach one establishment to one employee with multiple jobs

Attach :
![image](https://user-images.githubusercontent.com/77194719/222702478-3e9abd37-9a02-4fef-844e-4ab9e25b29f5.png)
Table :
![image](https://user-images.githubusercontent.com/77194719/222702572-9ea548f3-bce9-49ac-aded-ba0633fc9607.png)

But in table the establishment is duplicate, and i cannot attach many jobs at one time

**Reply by danharrin** [reply]
Send the form() code you used

**Reply by warsemann** [reply]
Form in HeaderActions ? : 

`                Tables\Actions\AttachAction::make()->preloadRecordSelect()
                ->form(fn ( Tables\Actions\AttachAction $action): array => [
                    $action->getRecordSelect(),
                    Forms\Components\Select::make('job_id')->options(Job::all()->pluck('name', 'id'))->required()->multiple(),
                ])`

**Reply by danharrin** [reply]
As I suggested, `$action->getRecordSelect()->multiple()`

**Reply by warsemann** [reply]
Multiple is ok, the front end is ok, th eproblem is the back end, can i override the attach method ? 

I need to attach establishment with employee (table employee_establishment) and  attach job with employee and establishment (table employee_job with pivot establishment_id)

Perhaps it s not the good database configuration, or should i use the associate method

**Reply by danharrin** [reply]
The job select should not be multiple() as you only store 1 job per row. Also make sure it's defined in the `withPivot()` of the belongstomany relationships. You shouldn't need to override the attach method

**Reply by warsemann** [reply]
But i would: 
Employee
       -->Establishment1
            -->Job1
            -->Job2
        -->Establishment2
            -->Job2    
            -->Job3

**Reply by danharrin** [reply]
You need a JSON column then

**Reply by warsemann** [reply]
Hi, i'm back :), i try another thing, can i attach a multiple select ? When i try the first id has been save but not the others

Thanks a lot !

**Comment by Traineratwot** [comment]
I have a mistake, tell me how to solve it
base model
```php
class Structure extends Model
{
    use SoftDeletes, HasFactory;

    ...

    public function logic(): BelongsToMany
    {
        return $this->belongsToMany(Logic::class,LogicStructure::class)->withPivot('available_values', 'default', 'description');
    }
}

```
Relation model
```php
class LogicStructure extends Model
{

    protected $casts = [
        'available_values' => 'array',
    ];
}


```

Attach with form

problem with "KeyValue"  Array to string conversion. Other field is worked

```php

AttachAction::make()
                    ->preloadRecordSelect()
                    ->form(function (AttachAction $action, Form $form): array {
                        return [
                            $action->getRecordSelect(),
                            KeyValue::make('available_values')->nullable(),
                            TextInput::make('default')->numeric()->nullable(),
                            MarkdownEditor::make('description')->nullable(),
                        ];
                    })

                ,
```

**Reply by gaetandezeiraud** [reply]
Same problem for me, have you find a solution?

**Reply by lanz1** [reply]
Same problem.
Seems that Filament doesn't cast the pivot correctly and passes the field as a string to the database.

Adding a mutator works for me:

`Tables\Actions\AttachAction::make()
...
->mutateFormDataUsing(function (array $data): array {
                        $data['multiselect_field_name'] = json_encode($data['multiselect_field_name']);
                        return $data;
                    })`
